firmware/spade/src/rpi/main.c

```diff
@@ -265,74 +320,185 @@ int main() {
-                     "                    \n"
-                     "                    \n"
-                     "                    \n"
-                     "                    \n"
-                     "                    \n"
-                     "    PLEASE UPLOAD   \n"
-                     "       A GAME       \n"
-                     "                    \n"
-                     "                    \n"
-                     "                    \n"
-                     "                    \n"
-                     "                    \n"
-                     " sprig.hackclub.com \n");
-    render_errorbuf();
-    st7735_fill_start();
-    render(st7735_fill_send);
-    st7735_fill_finish();
+    // Start a core to listen for keypresses.
+    multicore_reset_core1();
 
-    load_new_scripts();
-  }
+    update_save_version(); // init here to avoid irqs on other core
 
-  // Start a core to listen for keypresses.
-  multicore_launch_core1(core1_entry);
+    multicore_launch_core1(core1_entry);
 
-  /**
-   * We get a bunch of fake keypresses at startup, so we need to
-   * drain them from the FIFO queue.
-   *
-   * What really needs to be done here is to have button_init
-   * record when it starts so that we can ignore keypresses after
-   * that timestamp.
-   */
-  sleep_ms(50);
-  while (multicore_fifo_rvalid()) multicore_fifo_pop_blocking();
+    /**
+       * We get a bunch of fake keypresses at startup, so we need to
+       * drain them from the FIFO queue.
+       *
+       * What really needs to be done here is to have button_init
+       * record when it starts so that we can ignore keypresses after
+       * that timestamp.
+       */
+    sleep_ms(50);
+    while (multicore_fifo_rvalid()) multicore_fifo_pop_blocking();
 
-  /**
-   * Wait for a keypress to start the game.
-   *
-   * This is important so games with e.g. infinite loops don't
-   * brick the device as soon as they start up.
-   */
-  while(!multicore_fifo_rvalid()) {
-    strcpy(errorbuf, "                    \n"
-                     "                    \n"
-                     "                    \n"
-                     "                    \n"
-                     "                    \n"
-                     "                    \n"
-                     "                    \n"
-                     "  PRESS ANY BUTTON  \n"
-                     "       TO RUN       \n"
-                     "                    \n"
-                     "                    \n"
-                     "                    \n"
-                     "                    \n"
-                     "                    \n"
-                     " sprig.hackclub.com \n");
-    render_errorbuf();
-    st7735_fill_start();
-    render(st7735_fill_send);
-    st7735_fill_finish();
+    int games_i = 0;
 
-    load_new_scripts();
-  }
+    int games_len = 0;
+    Game* games = malloc(METADATA_MAX_ENTRIES * sizeof(Game));
```

Comment:
We might want to in future wrap all malloc's and frees in a #define - not sure if there is any precedent at all but it wouldn't be a bad idea